Lab 8 & 9 - JSON puppy and LevelDB

mohankrishna opened the pull request:
Fixes #103, Fixes #104 & Fixes #112

#### Changes proposed in this PR:

- Update project structure
- Add JSON tags to Puppy 
- Add Storer implementation with LevelDB
- Add command line flag `-d FILE` with long form `--data FILE` using [kingpin.v2](https://godoc.org/gopkg.in/alecthomas/kingpin.v2)
- A file containing JSON array of Puppies can be passed as an argument with `-d` or `--data` flag

Commit:
Create project structure

Commit:
Restructure code and add tests

Commit:
Add command line flag -d FILE with long form --data FILE using kingpin.v2

mohankrishna committed:
Merge branch 'master' into Lab_8

Commit:
Fix lint error

anzboi commented on 08_project/mohankrishna/pkg/mohankrishna-puppy/store/sync_store.go:
```diff
@@ -0,0 +1,55 @@
+)
+
+type SyncStore struct {
+	sm sync.Map
```
The sync store should be safe to use by multiple concurrent users. You should add a `mutex` to this class and wrap the necessary function with `mutex.Lock()` and `mutex.Unlock()`. See the relevant comments.

anzboi commented on 08_project/mohankrishna/pkg/mohankrishna-puppy/store/sync_store.go:
```diff
@@ -0,0 +1,55 @@
+	return &SyncStore{}
+}
+
+func (ss *SyncStore) CreatePuppy(puppy *types.Puppy) error {
```
The multiple store accesses here create a race condition.
Add the following lines to the start of this function (with the classes mutex)
```go
mutex.Lock()
defer mutex.Unlock()
```

anzboi commented on 08_project/mohankrishna/pkg/mohankrishna-puppy/store/sync_store.go:
```diff
@@ -0,0 +1,55 @@
+	return nil
+}
+
+func (ss *SyncStore) DeletePuppy(id uint32) error {
```
Same with this function.
```go
mutex.Lock()
defer mutex.Unlock()
```

anzboi commented on 08_project/mohankrishna/pkg/mohankrishna-puppy/store/storer_test.go:
```diff
@@ -0,0 +1,214 @@
+	suite.Run(t, &levelDBSuite)
+}
+
+func (s *storerSuite) TestMapStoreReadPuppySuccessful() {
```
Rename these functions to generic names since these are no longer `MapStore` specific.
eg: `TestStoreReadPuppySuccessful()`

anzboi commented on 08_project/mohankrishna/pkg/mohankrishna-puppy/store/leveldb_store.go:
```diff
@@ -0,0 +1,79 @@
+	ls.ldb.Close()
+}
+
+func (ls *LevelDBStore) CreatePuppy(puppy *types.Puppy) error {
```
Similar to my comments in sync store, this storer currently has race conditions. I'll leave it up to you to decide if you want this store to be thread-safe.

anzboi commented on 08_project/mohankrishna/pkg/mohankrishna-puppy/store/leveldb_store.go:
```diff
@@ -0,0 +1,79 @@
+	return ls.ldb.Delete(b, nil)
+}
+
+func (ls *LevelDBStore) GetAll() ([]*types.Puppy, error) {
```
Nice 👍

anzboi reviewed (commented):
Fantastic work so far.
Have a few comments regarding race conditions in two of your store implementations. sync store should be safe for concurrent use, I'll leave it up to you if you want leveldb to be so as well.
Other than that, tests and main function looks good.

Commit:
Add Mutex locking for sync store and address review comments

mohankrishna commented:
@anzboi Thanks for the review. I have updated SyncStore and LevelDB implementations with locking to prevent race conditions and renamed tests to generic names.

anzboi reviewed (approved):
LGTM 👍

anzdaddy commented on 08_project/mohankrishna/cmd/mohankrishna-puppy-server/main.go:
```diff
@@ -0,0 +1,54 @@
+	"io/ioutil"
+	"os"
+
+	types "github.com/anz-bank/go-training/08_project/mohankrishna/pkg/mohankrishna-puppy"
```
Better to have the package at `github.com/anz-bank/go-training/08_project/mohankrishna/pkg/puppy` and import it with the default name `puppy`.

anzdaddy commented on 08_project/mohankrishna/cmd/mohankrishna-puppy-server/main.go:
```diff
@@ -0,0 +1,54 @@
+	"os"
+
+	types "github.com/anz-bank/go-training/08_project/mohankrishna/pkg/mohankrishna-puppy"
+	store "github.com/anz-bank/go-training/08_project/mohankrishna/pkg/mohankrishna-puppy/store"
```
No need for the explicit local name.

anzdaddy commented on 08_project/mohankrishna/cmd/mohankrishna-puppy-server/main.go:
```diff
@@ -0,0 +1,54 @@
+	filePath := kingpin.Flag("data", "path to json data file").Short('d').String()
+	_, err := kingpin.CommandLine.Parse(args)
+	if err != nil {
+		fmt.Fprint(out, err)
```
Return `err` and do the printing in `main`.

anzdaddy commented on 08_project/mohankrishna/cmd/mohankrishna-puppy-server/main.go:
```diff
@@ -0,0 +1,54 @@
+
+var out io.Writer = os.Stdout
+
+func parseArgs(args []string) {
```
This function is misnamed. It does far more that parse arguments.

anzdaddy commented on 08_project/mohankrishna/cmd/mohankrishna-puppy-server/main_test.go:
```diff
@@ -0,0 +1,86 @@
+
+func TestParseArgsOutput(t *testing.T) {
+	// Given
+	r := require.New(t)
```
`r` is just testing support. It isn't part of the test's context. It should therefore appear before `// Given`.

anzdaddy commented on 08_project/mohankrishna/pkg/mohankrishna-puppy/store/leveldb_store.go:
```diff
@@ -0,0 +1,85 @@
+		return &types.Error{Code: types.ErrInvalidInput,
+			Message: fmt.Sprintf("The id:%d passed and the puppy's id:%d doesnt match", id, puppy.ID)}
+	}
+	b, _ := json.Marshal(id)
```
Handle errors. If they should be impossible for the given input, then `panic(err)` instead of `return err`.

anzdaddy commented on 08_project/mohankrishna/pkg/mohankrishna-puppy/store/leveldb_store.go:
```diff
@@ -0,0 +1,85 @@
+			Message: fmt.Sprintf("The id:%d passed and the puppy's id:%d doesnt match", id, puppy.ID)}
+	}
+	b, _ := json.Marshal(id)
+	bytes, _ := json.Marshal(puppy)
```
Handle errors. If they should be impossible for the given input, then `panic(err)` instead of `return err`.

anzdaddy commented on 08_project/mohankrishna/pkg/mohankrishna-puppy/store/leveldb_store.go:
```diff
@@ -0,0 +1,85 @@
+	ls.mutex.Lock()
+	defer ls.mutex.Unlock()
+	ok, _ := ls.ldb.Has(b, nil)
+	if !ok {
```
Fold into previous line.

anzdaddy commented on 08_project/mohankrishna/pkg/mohankrishna-puppy/store/leveldb_store.go:
```diff
@@ -0,0 +1,85 @@
+}
+
+func (ls *LevelDBStore) DeletePuppy(id uint32) error {
+	b, _ := json.Marshal(id)
```
Handle errors. If they should be impossible for the given input, then `panic(err)` instead of `return err`.

anzdaddy commented on 08_project/mohankrishna/pkg/mohankrishna-puppy/store/leveldb_store.go:
```diff
@@ -0,0 +1,85 @@
+	for iter.Next() {
+		var pup types.Puppy
+		err := json.Unmarshal(iter.Value(), &pup)
+		if err == nil {
```
Fold into previous line.

anzdaddy commented on 08_project/mohankrishna/pkg/mohankrishna-puppy/store/leveldb_store.go:
```diff
@@ -0,0 +1,85 @@
+			puppies = append(puppies, &pup)
+		}
+	}
+	iter.Release()
```
Should `defer iter.Release()` immediately after `iter := …`.

anzdaddy commented on 08_project/mohankrishna/pkg/mohankrishna-puppy/store/map_store.go:
```diff
@@ -0,0 +1,50 @@
+func (ms *MapStore) CreatePuppy(puppy *types.Puppy) error {
+	if _, exists := ms.m[puppy.ID]; exists {
+		return &types.Error{Code: types.ErrDuplicate,
+			Message: fmt.Sprintf("A puppy with ID: %d already exists", puppy.ID)}
```
Indent fields consistently. Move `Code: …` to its own line.

anzdaddy commented on 08_project/mohankrishna/pkg/mohankrishna-puppy/store/map_store.go:
```diff
@@ -0,0 +1,50 @@
+
+func (ms *MapStore) ReadPuppy(id uint32) (*types.Puppy, error) {
+	puppy, ok := ms.m[id]
+	if !ok {
```
Invert test and fold into previous line.

anzdaddy commented on 08_project/mohankrishna/pkg/mohankrishna-puppy/store/map_store.go:
```diff
@@ -0,0 +1,50 @@
+
+func (ms *MapStore) DeletePuppy(id uint32) error {
+	_, ok := ms.m[id]
+	if ok {
```
Fold into previous line.

anzdaddy commented on 08_project/mohankrishna/pkg/mohankrishna-puppy/store/map_store.go:
```diff
@@ -0,0 +1,50 @@
+func (ms *MapStore) UpdatePuppy(id uint32, puppy *types.Puppy) error {
+	if id != puppy.ID {
+		return &types.Error{Code: types.ErrInvalidInput,
+			Message: fmt.Sprintf("The id:%d passed and the puppy's id:%d doesnt match", id, puppy.ID)}
```
Indent fields consistently. Move `Code: …` to its own line.

anzdaddy commented on 08_project/mohankrishna/pkg/mohankrishna-puppy/store/storer_test.go:
```diff
@@ -0,0 +1,214 @@
+	"testing"
+
+	types "github.com/anz-bank/go-training/08_project/mohankrishna/pkg/mohankrishna-puppy"
+	tassert "github.com/stretchr/testify/assert"
```
Different local name not necessary

anzdaddy commented on 08_project/mohankrishna/pkg/mohankrishna-puppy/store/storer_test.go:
```diff
@@ -0,0 +1,214 @@
+)
+
+var (
+	puppy1 = func() types.Puppy {
```
Why vars instead of just regular functions?

anzdaddy commented on 08_project/mohankrishna/pkg/mohankrishna-puppy/store/storer_test.go:
```diff
@@ -0,0 +1,214 @@
+func (s *storerSuite) SetupSuite() {
+	switch s.impl {
+	case maps:
+		// create a map store
```
Comment states the obvious and is therefore unnecessary.

anzdaddy commented on 08_project/mohankrishna/pkg/mohankrishna-puppy/store/storer_test.go:
```diff
@@ -0,0 +1,214 @@
+		// create a map store
+		s.store = NewMapStore()
+	case smaps:
+		// create a sync map store
```
Comment states the obvious and is therefore unnecessary.

anzdaddy commented on 08_project/mohankrishna/pkg/mohankrishna-puppy/store/storer_test.go:
```diff
@@ -0,0 +1,214 @@
+		// create a sync map store
+		s.store = NewSyncStore()
+	case levelDB:
+		// create a level db store
```
Comment states the obvious and is therefore unnecessary.

anzdaddy commented on 08_project/mohankrishna/test/jsondata.json:
```diff
@@ -0,0 +1,5 @@
+    {"id":4662, "breed":"Sheep herder1", "colour":"Brown1", "value":1000},
+    {"id":4663, "breed":"Sheep herder2", "colour":"Brown2", "value":1000},
+    {"id":4664, "breed":"Sheep herder3", "colour":"Brown3", "value":1000}
+]
```
Missing final \n

anzdaddy commented on 08_project/mohankrishna/test/jsondata_currupted.json:
```diff
@@ -0,0 +1,5 @@
+    {"id":4662, "breed":"Sheep herder1", "colour":"Brown1", "value":1000},
+    {"id":4663, "breed":"Sheep herder2", "colour":"Brown2", "value":1000}
+    {"id":4664, "breed":"Sheep herder3", "colour":"Brown3", "value":1000}
+]
```
Missing final \n

mohankrishna commented:
@anzdaddy Thank you for reviewing the changes. Now that the PR is already merged how do i update the code with changes suggested? Shall i revert the changes with a new PR as github suggests or open a new PR with only the suggested changes?

camscale commented:
I would open a new PR with just the suggested changes. Once things are merged to master, there's no real going back. You can revert commits, but you should save that for when things have gone quite wrong.

Iteratively refining your work over multiple PRs works fine as a model for software development. The main thing to keep in mind is that ideally everything in the repo continues to work after every commit  - i.e. you don't split work over two commits where the repo is broken after the first commit. Doing this correctly means you can use `git bisect` to identify bugs as `git bisect` works best when you can checkout any commit and run the tests and those tests pass.
